### DISCUSSION TITLE
RVCop64 - UART usage (J2 connector) for networking...

### DISCUSSION THREAD

**Discussion by pottendo** [zeldin/OrangeCart#8] [RVCop64]
hi,

I'd like to make use of the 'Debug UART' - connector J2.
My use case:
connect some espressif ESP (e.g. ESP32 or 8266) via the J2 serial line (ideally HW UART on the ESP side) and talk serial to a modem SW, enabling a WiFi connection.
I have already some modem SW (see [here](https://github.com/pottendo/esp32-userport-driver)), so the ESP part would be easy. On Zephyr there's some 'ppp' driver, which may work as well. So with that network connectivity could be enabled for the RVCop64, opening some interesting options, IMHO.
I'd like to keep the console (printf on RISCV) to go out on the USB to be captured via litex-term; as I do today via a /dev/ttyACM0 device. If this isn't possible, also the C64 RVterm or some 'crossover' with litex-server would be an option... 

How's the bitstream to be configured wrt. the serial and/or usb options for this use case? I have to admit that the description [here](https://github.com/zeldin/RVCop64/blob/master/doc/debugging.md) is a bit confusing for me...

Thanks, pottendo

**Comment by zeldin** [comment]
I can understand that the debugging page is a bit daunting because there are so many options (like serial over JTAG, JTAG over serial etc).  And yet, your use case is not even included.  :smile:

If I understand you correctly, you want two UARTs available to the RiscV core, one connected to your `/dev/ttyACM0` and used for console I/O, and one connected to J2 to use for e.g. PPP.  Hardware wise this is not a problem at all, but having more than one UART isn't really supported by the `libbase` driver layer provided with LiteX, and that is why no such configuration can be selected with the provided configuration options.  That said, Zephyr should not have any issues with two UARTs, so all you need to do is make a small change to `hw/rtl/basesoc.py` to add the secondary UART.  After the call to `SoCCore.__init__`, you should add a call like

```
   self.add_uart("uart2")
```

This will add a second UART connected to J2 running at 115200 baud.  `add_uart` is defined in `hw/deps/litex/litex/soc/integration/soc.py`; you can check it out for additional arguments (like setting a different baudrate).

**Reply by pottendo** [reply]
hi Marcus, looks easy - thx!
I've built the bitstream already; the generated devicetree overlay looks weird though:
```
&uart0 {
    reg = <0xf0001000 0x4
        0xf0001004 0x4
        0xf0001008 0x4
        0xf000100c 0x4
        0xf0001010 0x4
        0xf0001014 0x4
        0xf0001018 0x4
        0xf000101c 0x4
        0xf0004800 0x4
        0xf0004804 0x4
        0xf0004808 0x4
        0xf000480c 0x4
        0xf0004810 0x4
        0xf0004814 0x4
        0xf0004818 0x4
        0xf000481c 0x4>;
    reg-names = "rxtx",
        "txfull",
        "rxempty",
        "ev_status",
        "ev_pending",
        "ev_enable",
        "tuning_word",
        "configured",
        "_rxtx",
        "_txfull",
        "_rxempty",
        "_ev_status",
        "_ev_pending",
        "_ev_enable",
        "_txempty",
        "_rxfull";
    interrupts = <0x0 0>;
};
```
The generated csr.h looks normal.
I assume litex's zephyr generator for device-tree overlays isn't perfect (yet). 
There are some example of multi-uart boards, so I'll continue to 'borrow' those examples... and connect it to ppp (one day).

Thank & happy hacking, pottendo

**Reply by zeldin** [reply]
Yes, it looks like you got both UARTs mapped to the same DT node.  `0xf00010xx` is the first UART, and `0xf00048xx` is the second one.  I'll take a look at the generator to see if I can figure out what is going on.

**Reply by zeldin** [reply]
Ok, I fixed the overlay generator.  Please note that the base device-tree now must contain a label `uart1` (the overlay will disable it if no second UART is enabled).

**Comment by zeldin** [comment]
For convenience, I just pushed an update that allows adding a second UART from the command line.  If you do e.g.
```
./bitstream.py --platform orangecart --uart usb_acm --uart2 serial
```
you should get your desired configuration.  The second UART will be mapped to address `f0004800` and IRQ 3.  It has the same register interface as the first one.

**Reply by pottendo** [reply]
... ok, will check this; thx, pottendo

**Reply by pottendo** [reply]
...before merging this into my tree I'd like to know if you plan to accept my pull request? (VexRiscV-SMP support).
bye, pottendo

**Reply by zeldin** [reply]
Sorry, haven't had time to look at it yet...

**Reply by pottendo** [reply]
fyi, I did the merge... needed to get fluent with git anyway. 
Went smooth, so expect another pull request soon
br, pottendo

**Reply by zeldin** [reply]
You don't need to make a new pull request, just keep pushing to the same branch and it gets added to the current PR.

**Comment by pottendo** [comment]
yes, didn't know that; thx.
The merged PR is now pushed to my branch/master. I've now added to automatically chose `--with-wishbone-memory' in case of the SMP CPU.
Now it's time get the second UART doing something... 

PS: the new dts generator looks good - my build systems seems to be different though (we've discussed/seen this some months ago already) - look at the registers:
```
&uart0 {
    reg = <0xf0001000 0x4
        0xf0001004 0x4
        0xf0001008 0x4
        0xf000100c 0x4
        0xf0001010 0x4
        0xf0001014 0x4
        0xf0001018 0x4
        0xf000101c 0x4>;
    reg-names = "rxtx",
        "txfull",
        "rxempty",
        "ev_status",
        "ev_pending",
        "ev_enable",
        "tuning_word",
        "configured";
    interrupts = <0x0 0>;
};
&uart1 {
    reg = <0xf0004800 0x4
        0xf0004804 0x4
        0xf0004808 0x4
        0xf000480c 0x4
        0xf0004810 0x4
        0xf0004814 0x4
        0xf0004818 0x4
        0xf000481c 0x4>;
    reg-names = "rxtx",
        "txfull",
        "rxempty",
        "ev_status",
        "ev_pending",
        "ev_enable",
        "txempty",
        "rxfull";
    interrupts = <0x3 0>;
};
```
I'll dig into this another time... I can overrule to the correct overlay in my zephyr branch anyway.
bye, pottendo

**Reply by zeldin** [reply]
I'm looking at the registers, but what exactly am I looking for?  Could you please be a bit more specific?  😕

**Reply by pottendo** [reply]
hi, the generator generates:
```
&uart0 {
[...]
        "tuning_word",
        "configured";
```
in the first uart - which won't fit to the Zephyr driver. uart1 is fine.

**Reply by zeldin** [reply]
That is because `uart0` in this case is a USB_ACM UART (from ValentyUSB).  The USB_ACM UART has a programmable baud rate (actually it hasn't, but baud rate in the common sense is not applicable so it lets you set any baudrate and it will still work).  `tuning_word` is the register used to change the baud rate.  If you configure the serial (J2) UART to have a programmable baud rate, it will also get a `tuning_word` register.  The `configured` register is specific to USB_ACM, and lets the software activate the UART more quickly (normally it self-activates after a delay).
So the generator is correct.  The last registers are different between the two UART implementations.  If the Zephyr driver wants to use those registers it needs to check whether the hardware implementation has them or not first.

**Reply by zeldin** [reply]
As far as I can tell though, the Zephyr driver only uses the registers `rxtx`, `txfull`, `rxempty`, `ev_pending` and `ev_enable`.  It has defines for `ev_status`, `txempty` and `rxfull`, but never use them.  So the driver should work fine with both implementations as it is now.

**Reply by pottendo** [reply]
ok, thanks for the explanation.

**Comment by pottendo** [comment]
Hi fellow-hackers,
here an update...
I've confirmed that my J2 (+my wobbly cable hack) works... by redirecting the Zephyr console to the second UART and collecting the output there via a FTDI/USB cable in another terminal program on my PC.

Now I found that the Zephyr litex-uart driver (also) can handly just _one_ uart device. This looks easy to extend (there are some other drivers, which handle more than one); so I'm on my way to connect the RVCop64 to the internet...
...stay tuned! 
pottendo

**Comment by pottendo** [comment]
hi,
I'v added support for 2 litex UARTS to my Zephyr branch, [here](https://github.com/pottendo/zephyr-pottendo)
All seems to work fine, except one case: Zephyr offers a 'shell-module' which gives kind of an interactive command-shell. This only can be used if one disables `SERIAL_SUPPORT_INTERRUPT` as there's a check for compatibility, which fails for some reason. This has probably nothing to do with the 2 uart bitstream. Unfortunately interrupt driven serial is a precondition to make the modem driver work... 
... so I decided to keep modem support, the shell isn't useful for me (yet).

I successfully even put the main console on the second uart residing on J2, confirming that also input (e.g. via the above shell) works.
Finally I also booted Linux with 2 UART support and confirmed a working bi-directional communication. There, full ppp network experience could be done as well... let's see if this fits into memory.

The (Zephyr-)PPP driver starts already talking its `AT` command on J2, so now it's time to connect something which reacts properly...
Exciting, imho!
br pottendo

**Comment by pottendo** [comment]
Hi fellow hackers,

the J2 connector talks to my ZiModem, however the Zephyr's gsm-ppp driver really insists to find a GSM modem there and tries to register in the network and several other GSM connection related commands. I suspect that attaching a real GSM module will work fine. Unfortunately I don't have one... :-(
ZiModem (which I've ported to an ESP32 in [this](https://github.com/pottendo/esp32-userport-driver) project) isn't too happy with the GSM stuff; I probably need to get a bunch of commands implemented there - even if those are no-ops. I'm not fluent in Hayes' "AT" speak, so I'm not sure if this is what I'd like to learn... 

Finally, I found an issue with the VexRiscV-smp soft-core on RVCop64. As written above, I need `SERIAL_SUPPORT_INTERRUPT` enabled on Zephyr to build gsm-driver. Understandable, as this is reactive to anything the uart may say at any time. 
Unfortunately **I don't see any uart interrupt arriving on this VexRiscV-smp** enabled bitstream. The console works on both uarts as expected, but those don't use interrupt driven communication, but polling (I assume).
Even on Linux, none of the uart is using interrupts, as far as I found out; the Linux liteuart driver isn't even registering an IRQ. Still I could exchange data via some terminal programs to my host PC in both directions on both uarts.
One difference between the two soft-cores is that the -SMP version defines a `PLIC` (platform local interrupt controller). However for basic usage on Zephyr (Timer, polling uarts, threading, etc.) this isn't even needed - one could omit the respective section in the device-tree.
I tried a lot to get the plic driver in Zephyr doing something, but it's just silent, as no interrupt are arriving there. 

Fortunately the VexRiscV (original RVCop64) CPU works as expected - with this configuration, I managed the modem tests as reported above. 
I tried to activate the 'intc' interrupt controller driver for the -SMP version, but this made Zephyr non-working... not even booting.

Any comments are welcome,
pottendo

**Reply by zeldin** [reply]
Right, so with a PLIC (which is presumably there precisely because of SMP -- the PLIC controls which hart handles which interrupt) you'll need to program the PLIC for an interrupt to reach the CPU.  But this also means that some driver needs to request the IRQ, otherwise the PLIC driver will never set it up and the interrupt will be discarded inside the PLIC.  Since you say that the driver does not request the IRQ, I'm guessing this is why you never see the interrupt arriving.  But that also means that even if the interrupt would arrive, the driver does nothing with it.  So I'm curious how, under those conditions, that things become working / not working depending on whether the interrupt is delivered?

**Reply by pottendo** [reply]
...the Zephyr driver does request the interrupt - sorry, if I was not precise.
See [here](https://github.com/pottendo/zephyr-pottendo/blob/main/drivers/interrupt_controller/intc_plic.c) around line 201.
It attaches the ISR and enables `RISCV_MACHINE_EXT_IRQ`, which is `11`.
It seems that the enabling the interrupt is wrong. I never saw the respective ISR being called. As you've mentioned, it may just not be connected to the right 'hart'.
I somehow came up with [this](https://github.com/pottendo/zephyr-pottendo/blob/main/soc/riscv/litex-vexriscv/irq.c) code; this made the timer (interrupt 7) work.
The `IRQ_CONNECT` macro works, as this generated the `_isr_sw_table` per side-effect. One can see that the ISRs are there in the right slot.

The Linux driver doesn't do this (as far as I see); this is what I meant.

bye, pottendo

**Comment by pottendo** [comment]
Hi @zeldin,

with some help from Zephyr & VexRiscV experts  
a) VexRiscV-smp improved to follow the recent FPU changes in Zephyr related to multi-threaded/context-switch issues (fixes also Linux!) and
b) Zephyr SMP (2 cores w/o FPU) running on the Orangecart. I've confirmed that the 2 cores improve speed in multithreaded applications: 77secs vs. 113 secs in my mandelbrot testcase. see below *)

ad a) fixes for vexriscv_smp are still to be pushed upstream, but @Dolu1990 confirmed that those will arrive soon

with b) In Zephyr the 'clint' (coming from the vexriscv-smp) sits at interrupt 3 (SoftIRQ) and 7 (timer) - these seems to be the default IRQ numbers also used in other RiscV systems, so I changed the irq for uart2 to 4 in basesoc.py
Using vexriscv-smp, I still don't see any interrupt 11 (external interrupt in Zephyr) arriving - this should be a PLIC generated irq, I assume...

When using vexriscv the external irq (riscv-intc driver) is arriving - this also is assigned irq 11 in Zephyr.
Can you tell me where the litex (or RVCop64) code defines the interrupt numbers for the FPGA? I'm obviously not clever enough to find how this connects... e.g. irq 11 is nowhere defined, but still works as external irq in vexriscv CPUs.
(probably my Python skills suck!)

Lastly: I hope to get my C64 back and maybe a more recent C64C on the next weekend - then i can do a more thorough test to find the right `phase_shift` value for best compatibility.

thanx, pottendo

*) as with vexriscv_smp the memory speed is far slower than with your vexriscv default choice (probably due to --with-wishbone-memory needed), even the dual-core vexriscv-smp version is significant slower than the single core vexriscv: 60sec vs. 77sec; with FPU support (vexriscv-smp, single core) things turn again: 2.8sec vs. 60sec ... ;-)
If you have an idea to improve the mem-speed also for vexriscv-smp it would be great. **Maybe you can have a look?!**
